### DISCUSSION TITLE
Use of Dialog (controlled) without Dialog.Trigger?

### DISCUSSION THREAD

**Discussion by janhoogeveen** [radix-ui/primitives#1234] [Help]
Hey everyone,

I'm trying to open a Dialog on route changes. The use case here is that I want a modal window on certain drilldown pages in Next.js using a catch-all route, and hide it when certain parameters are missing. I thought making a controlled Dialog would be a good solution here.

The way I open the Dialog is as follows:
1. I start at `/main/:mainId`
2. I navigate to `/main/:mainId/sub/:subId` though a `next/link`
3. Thanks to a `useEffect` hook listening to `:subId`, I set the Dialog to open.

```
<Dialog.Root open={isToolkitOpen}>
```

For the most part this is working as expected. There is one issue though. Once I close the Dialog, on touch devices things like sliders don't work anymore. They don't respond to any swipe events _until_ you tap on a non interactive part of the page. I suspect it's because Dialog tries to reset focus to the previous focus state and it can't find the Trigger. Or any trigger for that matter.

The docs say this about it.

**Escape key**
> Closes the dialog and moves focus to Dialog.Trigger.

I'm not using a Dialog.Trigger. Could this be the cause for this behaviour? It's not entirely clear for me if a Dialog implementation needs to use all parts listed in the documentation.

Update 1
Looks like the body has `pointer-events: none` set to it, which disappears as soon as I click or tap somewhere in the body.

Update 2
This is probably an issue due to the bodyPointerLock util, not because of the absence of a Trigger

**Comment by benoitgrelard** [comment]
Hey @janhoogeveen, generally you want to avoid showing dialogs without any user interactions, that's why the trigger exists as it is an integral part of the accessibility of the pattern (focus management, attributes, etc). I think that's generally a dark pattern UX-wise to open without user intent.

Regardless, in your case, if you don't use it, what you describe sounds like it could be a bug potentially.

Can you send me a quick sandbox reproducing it so I can take a look and see if it's actually a bug or just mis-use of the component?

Thanks 🙏

**Reply by janhoogeveen** [reply]
Yep can do!
I've also opened an issue describing this issue.
https://github.com/radix-ui/primitives/issues/1236

**Reply by janhoogeveen** [reply]
Oh yes, for context, the reason I don't use a Trigger is because we have tens of list items that could open a modal window, and I want that content to be visible on a new URL, not through some internal state. I thought you could only have one trigger attached to a dialog?

**Reply by benoitgrelard** [reply]
A bit tricky to visualise the situation exactly here, but there are things you can do architecture-wise when you want to have multiples of the same dialogs. Quite often, people try to re-use the same instance, instead of going towards component re-use.

If you can provide a simplified sandbox for this example, I'm happy to take a look.

**Reply by cdeutsch** [reply]
I'm also finding the need to do a "controlled" Dialog.

Current site has "url driven" dialogs, like @janhoogeveen is describing. Eg.

- List of items
- Click item link `/items/:itemId`
- Render Dialog with Overlay if `itemId` present

Accessibility wise, not sure "url driven" dialogs are a good idea, but that's how it's currently designed and programmed.

Not being able to directly link to an item's detail dialog would be a poor experience for a lot of users.

**Reply by jjenzz** [reply]
it sounds like your link could be the `Dialog.Trigger`?

```jsx
<Dialog.Trigger asChild><a href="/items/:itemId"></a></Dialog.Trigger>
```

or if that link takes you to a new page where the original link becomes unmounted, you can use the `open` and `onOpenChange` props to control the dialog on the new page.

**Reply by rohanrehman** [reply]
> Hey @janhoogeveen, generally you want to avoid showing dialogs without any user interactions, that's why the trigger exists as it is an integral part of the accessibility of the pattern (focus management, attributes, etc). I think that's generally a dark pattern UX-wise to open without user intent.
> 
> Regardless, in your case, if you don't use it, what you describe sounds like it could be a bug potentially.
> 
> Can you send me a quick sandbox reproducing it so I can take a look and see if it's actually a bug or just mis-use of the component?
> 
> Thanks 🙏

There are more user triggers than just buttons, or the button / intent was triggered on a different page.
If you'de like further details on use cases, Nexjs intercepting route in login modals is a great one!

**Comment by kylemh** [comment]
Hey @benoitgrelard you mentioned that dialogs that render without a trigger feel like a dark pattern, but what about `beforeunload` events, timer-based dialogs, or dialogs that render as a reaction to an API response?

There are countless ways that I've been asked to display a modal without a trigger. I feel like the API of Dialog is really limiting on this front.

**Reply by andy-hook** [reply]
I’m curious by “timer-based”, would an example be an advertising pop up? that’s a pretty clear dark pattern and while PMs often ask for these I think we all agree they are not great for UX or a11y, broadly speaking that's the point being made here (that something being shown without explicit intent from the user is not ideal)

In other cases I find it helpful to step back and view the interaction end to end, typically there is an “origin” of an action, an action would have to start somewhere even if it isn’t immediately obvious. For the async example, how was the request triggered? that surely has to originate from a user action.   Even if we end up controlling the dialog to achieve a UX, the trigger remains an important piece to communicate the “origin” of this interaction thus Radix knows how to restore focus and fill the a11y relationship.

I’m sure there are cases where a trigger is arguable, something like a command palette perhaps where it’s trendy to obfuscate behind a hot key. In that case it's important to remember the original context and restore focus on close in whatever logical way makes most sense for UX and a11y. In the vast majority of cases though a trigger is possible, it's simply not obvious / the UX of the feature may be questionable.

> I feel like the API of Dialog is really limiting on this front.

Radix will always attempt to align with UI best practice, as do we with the feedback in the channels. It's important to avoid being dogmatic though and I don't doubt there are some cases where a trigger is not possible. However, for the vast majority of uses a trigger can and should be used, and if it's not possible then the controlled props are there to help fill in the gaps for your own use case.

**Reply by kylemh** [reply]
> I’m curious by “timer-based”, would an example be an advertising pop up? that’s a pretty clear dark pattern and while PMs often ask for these I think we all agree they are not great for UX or a11y, broadly speaking that's the point being made here (that something being shown without explicit intent from the user is not ideal)

Agreed those are bad experiences; however, I think there are PLENTY of non-dark pattern dialogs which open asynchronously:
- Timed tests for educational platforms.
- Helper UI for users who have a slow internet connection or who lose internet connection (think about websockets for example).
- An "authentication expiring soon" dialog.

> In that case it's important to remember the original context and restore focus on close in whatever logical way makes most sense for UX and a11y. In the vast majority of cases though a trigger is possible, it's simply not obvious / the UX of the feature may be questionable.

I think this is what I'm getting at. I'm an engineer implementing features. I push back on designs that feel extremely inaccessible, but ultimately I want a component that helps be to be as responsible as possible. While the happy path (button triggerable) of Radix Dialog is a nice experience, I feel left in the dust with documentation for the times I need imperative control.

I don't want to just complain. I wanted to suggest some ideas:
- I wanted to suggest adjusting Dialog to be used with it's children as a render props exposing a `dismissDialog` method. This would help the experience for dealing with forms in dialogs whose form submissions are asynchronously and cannot simply be closed onclick.
- I also want to add examples about asynchronous form submissions in a Dialog and Dialogs that open in reaction to something async (not triggered via a button). By adding the examples, we can help users get these tasks done and remain as responsible as possible.

**Reply by jjenzz** [reply]
> I wanted to suggest adjusting Dialog to be used with it's children as a render props exposing a `dismissDialog` method. [..] dealing with forms in dialogs whose form submissions are asynchronously and cannot simply be closed onclick.

that's what the `open` prop is for. i agree it would be good to add an example similar to the following to the docs:

```jsx
const [open, setOpen] = React.useState(false);

<Dialog.Root open={open} onOpenChange={setOpen}>
  <Dialog.Trigger />
  <Dialog.Content>
   <form 
    onSubmit={event = {
      doSomeAsyncThing().then(() => setOpen(false));
      event.preventDefault();
    }}
   >
    {/* ... */}
    <Button type="submit">Submit</Button>    
   </form>
  </Dialog.Content>
</Dialog.Root>
```

for dialogs that should open without a trigger, i wonder if the `Dialog` could log a warning if there is no `Dialog.Trigger` that instructs user to pass a `restoreFocusElement` prop to the dialog? it would accept a node and would silence the warning. @benoitgrelard @andy-hook thoughts?

that prop would also be really useful for the `DropdownMenu` / `Dialog` composition. the dialog could log this warning if it cannot find the trigger when it tries to restore focus (e.g. it was unmounted because `DropdownMenu` closed when dialog opened).

**Reply by andy-hook** [reply]
@kylemh Cheers for expanding on the “timer-based“ uses, 100% agree that more examples would help folks. We even worked on a “Recipes” concept recently as part of an internal hackweek because we feel there’s a missing resource to solve these types of questions. e.g. dialog on async form https://codesandbox.io/s/form-data-async-dialog-ntqgpe?file=/src/App.js

> i wonder if the Dialog could log a warning if there is no Dialog.Trigger that instructs user to pass a restoreFocusElement prop to the dialog? it would accept a node and would silence the warning.

@jjenzz Yeah something like this could be helpful.

**Reply by kylemh** [reply]
I ended up at that solution jen outlined, but I was essentially worried that I was doing something inaccessible. Even if I was confident it was accessible, it feels really confusing to have both explicit (code-written) and implicit (internal Dialog state) controls. I made an abstraction where it's either-or. `Dialog.Trigger` nor `Dialog.Close` are not exposed, so it's all driven by props and the render prop for dismissing that I mentioned before. I guess - even here - there's still some implicitness (overlay clicks and hitting escape key).

At the very least it sounds like y'all are okay with another example for Dialog docs regarding async form submission in-Dialog. Mind if I contribute that?

**Reply by andy-hook** [reply]
> At the very least it sounds like y'all are okay with another example for Dialog docs regarding async form submission in-Dialog. Mind if I contribute that?

Absolutely! 🙏

**Reply by benoitgrelard** [reply]
@jjenzz I don't like the idea of the `restoreFocusElement` as it doubles what `onCloseAutoFocus` does already.

**Reply by jjenzz** [reply]
@benoitgrelard fair, same idea applies tho, just log a warning that tells them to implement `onCloseAutoFocus` instead.

**Comment by cuginoAle** [comment]
Hi, I am having a similar issue. I need to show a confirm box if the user tries to navigate away from a page containing an unsaved form.
On route change I set my state var `show` to true and the modal pops-up. On close I would expect the focus to be restored to the navigation item the user clicked on.. but this doesn't seem to be the case.
Is this because the modal doesn't contain a `Trigger`?
IS there any way to work around this?

cheers

**Reply by benoitgrelard** [reply]
You can/should control focus manually. You do this by preventing the default behaviour in `onCloseAutoFocus` on `Content`:

```tsx
onCloseAutoFocus={event => {
  event.preventDefault();
  somethingToFocusRef.current.focus();
}}
```

**Reply by cuginoAle** [reply]
Ah... that's a shame. Why should I control the focus manually anyway?
Since the modal is handling the focus I would expect it to restore the focus on close. This feature is available on many other ui libs.
Is there any technical reason not to implement this behaviour?
thanks

**Reply by benoitgrelard** [reply]
Just to clarify if you use the `Trigger` part, it would return automatically to the opening trigger.
We offer customisation via `onOpenAutoFocus` and `onCloseAutoFocus` which lets you prevent the built-in behaviour and run your own.

**Reply by cuginoAle** [reply]
Yes, thanks Benoit.
Since restoring the focus is a very frequent requirement (imho) I think it would be great if the modal could have this feature already built in.
Thanks  for the amazing work anyway :)

**Reply by benoitgrelard** [reply]
>  I think it would be great if the modal could have this feature already built in.

I feel like we're getting confused here, it is built-in, that's what I was saying above. By default, on open, focus will be moved to the first interactive element and on close, returned to the trigger.

**Reply by kylemh** [reply]
What if one uses `window.beforeunload`? In that situation, wouldn't you not want focus to return to the trigger but rather the previously focused element?

**Reply by cuginoAle** [reply]
I got that, what I mean is that I shouldn't be forced to use the Trigger to get the focus restored. 
Say I have ten possible CTAs to trigger a modal dialog (in my case a nav menu and other links in the footer); in this scenario I am not sure how to write some clean and maintainable code (using the Trigger) to implement a Modal dialog that would restore the focus on close.
On the other hand, if I use the `onCloseAutoFocus` callback then I need to manually keep track of `whateverWasFocusedBeforeRef` to restore the focus... which is something the lib could do for me.
Is focus restoration the only reason to have the Trigger?

**Reply by benoitgrelard** [reply]
> Is focus restoration the only reason to have the Trigger?

It isn't the only reason, the trigger also contains other accessibility attributes like `aria-haspopup`, `aria-expanded` and `aria-controls`.

> I am not sure how to write some clean and maintainable code (using the Trigger) to implement a Modal dialog that would restore the focus on close.

It's a case of creating a reusable component, rather than trying to reuse a single instance. These are old demos, but take a look at them. You could either:

- reuse the whole dialog: https://codesandbox.io/s/reused-dialog-lwmpj
- reuse the content if you need different triggers: https://codesandbox.io/s/reused-dialogcontent-0oikl

But the possibilities are really endless given they are just React components all the way down.

Hope this help!

**Reply by cuginoAle** [reply]
Thanks for taking the time to reply and provide examples.
I am still not convinced the API provides the best dev experience but I am sure this is something that has already been discussed between you maintainers.

Once again thanks and keep up with the good work!

**Reply by nemethricsi** [reply]
@benoitgrelard What is with the situation when I need to use a controlled dialog? I have no trigger exposed. Can I get focus back somehow after closing the dialog?

**Comment by NazarK0** [comment]
Hi there! I`m having issue with Dialog and Trigger too. In my form I have control panel with saving options, and on click  any of this option should open a dialog with progress bar. But it only should appear if form valid, and validation occur also on click this button. Could anyone show how to solve this issue?

**Comment by SoftChunck** [comment]
Its simple, **Dialog** require trigger right, _pass a button as a trigger_ and give button some id like `"alert"`, hide the button from ui using CSS, and then when ever you want to trigger an Dialog through a click  #1678  by code like, `document.getElementById('alert').click()`

**Reply by kylemh** [reply]
Not only is that not simple (very confusing for a dev to come on and figure all that out from existing code), but it's inaccessible for people that use assistive technology and against the spirit of this library.

**Comment by jalooc** [comment]
Another case in which having this api does not hold up is opening a modal from a popover menu. I.e. you have a button that opens a menu of which an item's click triggers opening a modal. You usually want the menu to hide upon an item click, but if we hide (unmount) that menu, the trigger disappears and so does the entire modal. The only way to solve this with the current api is probably using some hacks around keeping the modal rendered, but not visible, which is pretty dirty.

**Reply by jjenzz** [reply]
there are examples of how to handle that here https://github.com/radix-ui/primitives/issues/1836

**Reply by jalooc** [reply]
Thanks for that link @jjenzz , it's got some nice patterns 👍 But I'm afraid it does not really solve the overal problem, only presents particular workarounds for particular aspects of the problem. Applying those patterns would either/and:

1. hinder custom modal reusability (if I have a custom modal that I reuse in other places, I can't reuse it with some of the suggested patterns cause the part of the modal's logic lays inside the menu logic);
2. make it impossible to have different dialog root logic for each modal;
3. hinder atomicity and composability (the suggested patterns make the menu tightly coupled with modal - what if I don't want to use any modal in some instances of the menu).

All the above points are just one, not uncommon real life scenario: A composable popover menu whose items passed through props that can trigger arbitrary logic, a dialog opening being only one of the possibilities.

---

As much as I admire Radix, looking at the number of problems that devs encounter with the Dialog's api and the multitude of workarounds to overcome them depending on particular case, I think the Dialog api is just a pain to work with, making otherwise basic logic really hard to achieve.

I realize that the Dialog's api shape is dictated by accessibility, but can't the Dialog's architecture be based a global manager for opened modals (probably like in most/all other libraries), **and** have the accessible trigger elements at the same time? I don't think these are exclusive ideas.

I also realize that collocation is behind the choice, but then - _modals_ in general **are** global beings, so, while I'm a huge defender of collocation myself, imho modals management is a very natural case for a little bit of global state. (although if I understood [your article](https://jjenzz.com/avoid-global-state-colocate/) correctly, you'd still consider this kind of _global state_ collocated to the Dialog component, if archtectured properly, so even better 🙂)

**Comment by CodeSmith32** [comment]
I have yet another situation; I'm hoping the solution's already out there:

**I have a modal window that opens by some normal button, etc. Somewhere on this modal is a button.<br>When clicked, the current modal window closes, and a *different* modal opens.**

In this case, similar to the Dropdown Item, there is a clear trigger that's inside a component that disappears once triggered. Only, in this case, it's in *another modal*.

Following the same logic in #1836 for the Dropdown, one might think to put the second dialog's Trigger inside the first dialog's Content and then wrap the first dialog inside the second. But if I'm guessing correctly, sticking a trigger inside of a dialog's content will link to *that* dialog, not one that is wrapping it. Is that correct?

e.g.

```jsx
<Dialog.Root> {/* second dialog */}

  <Dialog.Root> {/* first dialog*/}

    <Dialog.Trigger> {/* open first dialog */}
      <button>Open first dialog</button>
    </Dialog.Trigger>

    <Dialog.Portal>
      <Dialog.Overlay />
      <Dialog.Content>

        <p>First dialog content</p>

        <Dialog.Trigger> {/* This trigger links to the first dialog, not the second, doesn't it? */}
          <Dialog.Close>
            <button>Open second dialog</button>
          </Dialog.Close>
        </Dialog.Trigger>

      </Dialog.Content>
    </Dialog.Portal>
  </Dialog.Root>

  <Dialog.Portal>
    <Dialog.Overlay />
    <Dialog.Content>
      <p>Second dialog content</p>
    </Dialog.Content>
  </Dialog.Portal>

</Dialog.Root>
```

If this is the case, how could this situation be handled correctly?

**Reply by CodeSmith32** [reply]
Or should I just disable the trigger?

It sounds like everyone's recommending against this, but it appears the Trigger can simply be disabled by hiding it. Then the dialog can be opened programmatically with `open` and `onOpenChange`.

```jsx
<Dialog.Trigger style={{display: "none"}} />

// or with Tailwind:
<Dialog.Trigger className="hidden" />
```

For my situation, this leads to the following possible solution:

```jsx
const [open, setOpen] = useState(false);

return <>
  {/* first dialog*/}
  <Dialog.Root>
    <Dialog.Trigger> {/* open first dialog */}
      <button>Open first dialog</button>
    </Dialog.Trigger>

    <Dialog.Portal>
      <Dialog.Overlay />
      <Dialog.Content>

        <p>First dialog content</p>

        <Dialog.Close>
          <button onClick={() => setOpen(true)}>
            Open second dialog
          </button>
        </Dialog.Close>

      </Dialog.Content>
    </Dialog.Portal>
  </Dialog.Root>

  {/* second dialog */}
  <Dialog.Root open={open} onOpenChange={setOpen}>
    {/* hide second dialog's trigger */}
    <Dialog.Trigger style={{display: "none"}} />

    <Dialog.Portal>
      <Dialog.Overlay />
      <Dialog.Content>

        <p>Second dialog content</p>

      </Dialog.Content>
    </Dialog.Portal>
  </Dialog.Root>
</>
```

But again, I'd lean toward a more *recommended* solution if there is one..

**Comment by reillo** [comment]
I resorted to use different library such as https://headlessui.com/ for the dialog.

**Reply by rohanrehman** [reply]
Thanks for sharing the team there is quite talented!

**Comment by 8spw** [comment]
only one thing that comes in my mind is to make custom `ref` and apply it to `Dialog.Close` I know its really bad practice but it works pretty well

```js
<DialogClose ref={closeDialogRef} className="hidden" />

closeDialogRef.current?.click();
```

**Reply by kluplau** [reply]
Another approach, quite similar to this, is using [useImperativeHandle](https://react.dev/reference/react/useImperativeHandle).

In your Dialog component: 

```tsx

export interface DialogRef {
  open: () => void
  close: () => void
}

export const Dialog = forwardRef<DialogRef, Props>(
  (props, ref) => {
    const [open, setOpen] = useState(false)

    useImperativeHandle(ref, () => ({
      open: () => setOpen(true),
      close: () => setOpen(false),
    }))

    return (
      <RadixDialog.Root open={open} onOpenChange={(open) => setOpen(open)}>
      ....
```

Where you use your Dialog component
```tsx
  const dialog = useRef<DialogRef>(null)

  dialog.current?.open() // Call this any where in the scope where the useRef is defined.

  return (
    <Dialog
      ref={dialog}
      ...
```

**Comment by topovik** [comment]
`<button onClick={(e) => {
e.preventDefault;
e,stopPropagation();
setOpen(false);
}}>Close Dialog</button>`
